compiler/src/iree/compiler/Codegen/Dialect/GPU/Transforms/DistributeMmaToLanes.cpp

```diff
@@ -0,0 +1,150 @@
+  using OpInterfaceRewritePattern::OpInterfaceRewritePattern;
+  LogicalResult matchAndRewrite(linalg::LinalgOp linalgOp,
+                                PatternRewriter &rewriter) const override {
+    auto loweringConfig =
+        getLoweringConfig<IREE::GPU::LoweringConfigAttr>(linalgOp);
+    if (!loweringConfig) {
+      return failure();
+    }
+    IREE::GPU::MmaInterfaceAttr kind = loweringConfig.getMmaKind();
+    if (!kind) {
+      return failure();
+    }
+    if (failed(convertContractionToMultiMma(rewriter, linalgOp, kind))) {
+      return failure();
+    }
+    return success();
+  }
+};
+
+LogicalResult fuseProducersGreedily(RewriterBase &rewriter,
+                                    scf::ForallOp laneForall) {
+
+  std::deque<tensor::ExtractSliceOp> candidates;
+  laneForall->walk([&](tensor::ExtractSliceOp extractSliceOp) {
+    auto producer = extractSliceOp.getSource().getDefiningOp<TilingInterface>();
+    if (producer && producer->getBlock() != laneForall.getBody()) {
+      candidates.push_back(extractSliceOp);
+    }
+  });
+
+  SmallVector<LoopLikeOpInterface> loops = {laneForall};
+
+  OpBuilder::InsertionGuard g(rewriter);
+  while (!candidates.empty()) {
+    // Traverse the slices in BFS fashion.
+    tensor::ExtractSliceOp candidateSliceOp = candidates.front();
+    candidates.pop_front();
+
+    // Materialize the slice of the producer in place.
+    std::optional<scf::SCFFuseProducerOfSliceResult> fusedProducer =
+        scf::tileAndFuseProducerOfSlice(rewriter, candidateSliceOp, loops);
+    if (!fusedProducer)
+      continue;
+
+    // We have no way to know whether a multi-use value can be yielded from the
+    // parallel loop so never yield a replacement.
+
+    // Add more fusion candidates to the worklist.
+    for (auto tiledOp : fusedProducer->tiledOps) {
+      for (OpOperand &operand : tiledOp->getOpOperands()) {
+        auto sliceOp = operand.get().getDefiningOp<tensor::ExtractSliceOp>();
+        if (!sliceOp)
+          continue;
+        candidates.push_back(sliceOp);
+      }
+    }
+  }
+  return success();
+}
+
+void DistributeMmaToLanesPass::runOnOperation() {
+  MLIRContext *context = &getContext();
+  auto funcOp = getOperation();
+
+  // Step 1. Convert configured linalg ops to multi_mma.
+  {
+    RewritePatternSet patterns(context);
+    patterns.add<ConvertToMultiMma>(context);
+    if (failed(applyPatternsAndFoldGreedily(funcOp, std::move(patterns)))) {
+      funcOp.emitError() << "failed to convert linalg to multi_mma";
+      return signalPassFailure();
+    }
+  }
+
+  // Step 2. Distribute multi_mma ops to lanes and greedily fuse producers.
+  SmallVector<IREE::GPU::MultiMmaOp> mmaOps;
+  funcOp.walk([&](IREE::GPU::MultiMmaOp mmaOp) { mmaOps.push_back(mmaOp); });
+  IRRewriter rewriter(funcOp);
+  for (auto mmaOp : mmaOps) {
+    rewriter.setInsertionPoint(mmaOp);
+    FailureOr<scf::ForallOp> maybeLaneForall =
+        distributeMultiMmaOp(rewriter, mmaOp);
```

Comment:
Distribute is accurate here because it isn't doing tiling. It is going from the subgroup representation directly to the thread representation, so it's not looking at iteration spaces, it's just extracting hard coded slices of the inputs based on the mma kind.